File: 08pairProduct.js





// using a hashmap (object)

// n = length of array
// time:  O(n)
// space: O(n)

const pairProduct = (numbers, targetProduct) => {
  const previousNums = {};
  for (let i = 0; i < numbers.length; i += 1) {
    const num = numbers[i];
    const complement = targetProduct / num;
    
    if (complement in previousNums) return [ previousNums[complement], i ];
    
    previousNums[num] = i;
  }
};

File: Python/08pairProduct.py

# n = length of numbers list
# Time: O(n)
# Space: O(n)

# using a hashmap(dictionary)


def pair_product(numbers, target_product):
  previous_nums = {}

  for index, num in enumerate(numbers):
    complement = target_product / num

    if complement in previous_nums:
      return (index, previous_nums[complement])

    previous_nums[num] = index



